neural_speed/application/main_pybind.cpp

```diff
@@ -54,6 +56,252 @@
+    ret_seq.n_total = 0;
+    ret_seq.gen_conf.max_new_tokens = params.n_predict;
+    ret_seq.gen_conf.min_new_tokens = params.min_new_tokens;
+    ret_seq.gen_conf.length_penalty = params.length_penalty;
+    ret_seq.gen_conf.do_early_stopping = params.do_early_stopping;
+    ret_seq.query_id = query.id;
+    return ret_seq;
+  }
+
+  Query Sequence2Query(const sequence& seq) {
+    Query ret_query;
+    ret_query.id = seq.query_id;
+    int ret_size = return_prompt ? seq.prompt_ids.size() + seq.generated_ids.size() : seq.generated_ids.size();
+    ret_query.token_ids.resize(ret_size);
+    if (return_prompt) {
+      std::copy(seq.prompt_ids.cbegin(), seq.prompt_ids.cend(), ret_query.token_ids.begin());
+      std::copy(seq.generated_ids.cbegin(), seq.generated_ids.cend(),
+                ret_query.token_ids.begin() + seq.prompt_ids.size());
+    } else {
+      std::copy(seq.generated_ids.cbegin(), seq.generated_ids.cend(), ret_query.token_ids.begin());
+    }
+    return ret_query;
+  }
+
+  void InitServerParams(const std::string& model_path, int max_new_tokens, int n_batch, int ctx_size, int seed,
+                        int threads, float repetition_penalty, int num_beams, bool do_sample, int top_k, float top_p,
+                        float temperature, int min_new_tokens, float length_penalty, bool early_stopping, int n_keep,
+                        int n_discard, bool shift_roped_k, int batch_size, model_vocab::id pad_token,
+                        const std::string& memory_dtype, const bool& continuous_batching, const int& max_request_num,
+                        const float& model_scratch_enlarge_scale) {
+    init_gpt_params(&params, model_path, max_new_tokens, n_batch, ctx_size, seed, threads, repetition_penalty,
+                    num_beams, do_sample, top_k, top_p, temperature, min_new_tokens, length_penalty, early_stopping,
+                    n_keep, n_discard, shift_roped_k, batch_size, pad_token, memory_dtype, continuous_batching,
+                    max_request_num, model_scratch_enlarge_scale);
+  }
+
+  ~ModelServer() {
+    // "synchronized" function
+    // stop spinning after calling ResponseCallback for the last query
+    py::print("Stopping CPP server...");
+    running = false;
+    {
+      py::gil_scoped_release releaser;
+      worker.join();
+    }
+    py::print("CPP server stopped!");
+  }
+
+ private:
+  // response function from outside for collecting generation results and checking server working status
+  const ResponseCallback response;
+  // waiting pool for new queries added into server
+  std::vector<Query> waiting;
+  // lock for waiting pool
+  std::mutex queue_mtx;
+  // status for telling server if it still need to continue running or not
+  // true: checking waiting pool and performing one step (or waiting new query)
+  // false: stop server
+  bool running;
+  gpt_params params;
+  // server policy (only FCFS (first come, first serve) now)
+  std::string policy;
```

Comment:
why not enum here?

Reply:
added enum [here](https://github.com/intel/neural-speed/blob/cont_batching/neural_speed/models/model_utils/pool.h#L36-L39). And let the user pass a string may be easier than pass a pybind enum class?